src/ciphersuite/mod.rs

```diff
@@ -201,84 +202,223 @@ impl KdfLabel {
+        // and we can't recover from it.
+        assert!(
+            self.mls_version == ikm.mls_version,
+            "{} != {}",
+            self.mls_version,
+            ikm.mls_version
+        );
+        assert!(
+            self.ciphersuite == ikm.ciphersuite,
+            "{} != {}",
+            self.ciphersuite,
+            ikm.ciphersuite
+        );
+
+        Self {
+            value: hkdf_extract(
+                self.ciphersuite.hmac,
+                self.value.as_slice(),
+                ikm.value.as_slice(),
+            ),
+            mls_version: self.mls_version,
+            ciphersuite: self.ciphersuite,
+        }
+    }
+
+    /// HKDF expand where `self` is `prk`.
+    pub(crate) fn hkdf_expand(&self, info: &[u8], okm_len: usize) -> Result<Self, HKDFError> {
+        let key = hkdf_expand(self.ciphersuite.hmac, &self.value, info, okm_len);
+        if key.is_empty() {
+            return Err(HKDFError::InvalidLength);
+        }
+        Ok(Self {
+            value: key,
+            mls_version: self.mls_version,
+            ciphersuite: self.ciphersuite,
+        })
+    }
+
     /// Expand a `Secret` to a new `Secret` of length `length` including a
     /// `label` and a `context`.
-    pub fn kdf_expand_label(
-        &self,
-        ciphersuite: &Ciphersuite,
-        label: &str,
-        context: &[u8],
-        length: usize,
-    ) -> Secret {
-        let info = KdfLabel::serialized_label(context, label, length);
+    pub(crate) fn kdf_expand_label(&self, label: &str, context: &[u8], length: usize) -> Secret {
+        let full_label = format!("{} {}", self.mls_version, label);
         log::trace!(
             "KDF expand with label \"{}\" and {:?} with context {:x?}",
-            label,
-            ciphersuite.name(),
+            &full_label,
+            self.ciphersuite.name(),
             context
         );
+        let info = KdfLabel::serialized_label(context, full_label, length);
         log::trace!("  serialized context: {:x?}", info);
         log_crypto!(trace, "  secret: {:x?}", self.value);
-        ciphersuite.hkdf_expand(self, &info, length).unwrap()
+        self.hkdf_expand(&info, length).unwrap()
     }
 
     /// Derive a new `Secret` from the this one by expanding it with the given
     /// `label` and an empty `context`.
-    pub fn derive_secret(&self, ciphersuite: &Ciphersuite, label: &str) -> Secret {
+    pub(crate) fn derive_secret(&self, label: &str) -> Secret {
         log_crypto!(
             trace,
             "derive secret from {:x?} with label {} and {:?}",
             self.value,
             label,
-            ciphersuite.name()
+            self.ciphersuite.name()
         );
-        self.kdf_expand_label(ciphersuite, label, &[], ciphersuite.hash_length())
+        self.kdf_expand_label(label, &[], self.ciphersuite.hash_length())
+    }
+
+    /// Update the ciphersuite and MLS version of this secret.
```

Comment:
I think this is a valuable observation that would be nice to have in the right place. Is there an existing issue to document this in? Alternatively, create a new one?

Reply:
See #383